Make sure to generate coverage base reports before all tests

knorth55 opened the pull request:
Now cpp base reports are generated after cpp tests, which dependencies are not correctly set.
I changed to depend the base report generation on `tests` target to make sure that all base reports are generated before all the tests.
I also make `cleanup coverage` targets to run before base report generation.

so the dependency order is 
`cleanup coverage` -> `create base report` -> `run tests` -> `create coverage report`

I also refactored cmake codes to
- suppress print
- add comments on cmake log
- remove dependencies to avoid cyclic dependency.

knorth55 commented:
Now cpp packages like `gitai_druid`'s coverage is too low, and this PR fixes the issue.

garaemon commented on cmake/Modules/CodeCoverage.cmake:
```diff
@@ -362,14 +353,13 @@ function(ADD_CODE_COVERAGE)
                 || echo "WARNING: No cpp report to output"
         # add baseline counters
-        COMMAND ${LCOV_PATH} -a ${PROJECT_BINARY_DIR}/${Coverage_NAME}.base -a ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info
+        COMMAND ${LCOV_PATH} -a ${PROJECT_BINARY_DIR}/${Coverage_NAME}_base_cpp.info -a ${PROJECT_BINARY_DIR}/${Coverage_NAME}.info
```
Please add it to `OUTPUT`.
I think some other files are missing from `OUTPUT`.

knorth55 replied:
I change to use `add_custom_command` and `add_cutom_target` to add `output`.

garaemon reviewed (commented):
The dependency graph I understand is like this
```mermaid
flowchart TD
    _run_tests_py("_run_tests_${PROJECT_NAME}_python_base_coverage_report") --> tests
    run_tests_py("run_tests_${PROJECT_NAME}_python_base_coverage_report") --> tests
    _run_tests_cpp("_run_tests_${PROJECT_NAME}_cpp_base_coverage_report") --> tests
    run_tests_cpp("run_tests_${PROJECT_NAME}_cpp_base_coverage_report") --> tests
    cleanup_cpp("${Coverage_NAME}_cleanup_cpp") --> _run_tests_cpp 
    cleanup_py("${Coverage_NAME}_cleanup_py") --> _run_tests_py 
```

Given that graph, I assume that `${Coverage_NAME}_cleanup_py` and `run_tests_${PROJECT_NAME}_python_base_coverage_report` can run at the same time.
How do you prevent it? Could you modify the mermaid diagram?

knorth55 commented:
@garaemon that's true... let me check.

knorth55 commented:
It is not clearly written, but it seems `run_tests_${PROJECT_NAME}` target for `catkin_make` like `catkin_make run_tests_gitai_tools`.
so I removed `run_tests_${PROJECT_NAME}` target because we use `catkin tools`.

I also refactored the dependency and now the graph is like this

knorth55 commented:
Now the flow is like this.
I change to skip all coverage report generation when there is no test in the repo.

### For package with tests

```mermaid
flowchart TD
    clean_test_results --> build
    build -->  cleanup_py
    build -->  cleanup_cpp
    _run_tests_python_base_coverage_report_command("${PROJECT_NAME}_base_python.xml") --> _run_tests_python_base_coverage_report
    _run_tests_python_base_coverage_report("_run_tests_${PROJECT_NAME}_python_base_coverage_report") --> tests
    cleanup_py("${Coverage_NAME}_cleanup_py") --> _run_tests_python_base_coverage_report_command
    _run_tests_cpp_base_coverage_report_command("${PROJECT_NAME}_base_cpp.info") --> _run_tests_cpp_base_coverage_report
    _run_tests_cpp_base_coverage_report("_run_tests_${PROJECT_NAME}_cpp_base_coverage_report") --> tests
    cleanup_cpp("${Coverage_NAME}_cleanup_cpp") -->  _run_tests_cpp_base_coverage_report_command
    tests --> _run_tests_project("_run_tests_${PROJECT_NAME}")
    _run_tests_project --> _run_tests_cpp_coverage_report("${PROJECT_BINARY_DIR}/${Coverage_NAME}_cpp.info")
    _run_tests_project --> _run_tests_python_nosetests_coverage_report("${PROJECT_BINARY_DIR}/${Coverage_NAME}_nosetests_python.xml")
    _run_tests_project --> _run_tests_python_pytests_coverage_report("${PROJECT_BINARY_DIR}/${Coverage_NAME}_pytests_python.xml")
    _run_tests_python_nosetests_coverage_report --> _run_tests_python_coverage_report("${PROJECT_BINARY_DIR}/${Coverage_NAME}_python.xml")
    _run_tests_cpp_coverage_report --> coverage_report("${Coverage_NAME}_coverage_report") 
    _run_tests_python_coverage_report --> coverage_report
    _run_tests_python_pytests_coverage_report --> coverage_report
```

### For package without tests

```mermaid
flowchart TD
    _run_tests_python_base_coverage_report_command("${PROJECT_NAME}_base_python.xml") --> _run_tests_python_base_coverage_report
    _run_tests_python_base_coverage_report("_run_tests_${PROJECT_NAME}_python_base_coverage_report") --> _run_tests_project("_run_tests_${PROJECT_NAME}")
    cleanup_py("${Coverage_NAME}_cleanup_py") --> _run_tests_python_base_coverage_report_command
    _run_tests_cpp_base_coverage_report_command("${PROJECT_NAME}_base_cpp.info") --> _run_tests_cpp_base_coverage_report
    _run_tests_cpp_base_coverage_report("_run_tests_${PROJECT_NAME}_cpp_base_coverage_report") --> _run_tests_project("_run_tests_${PROJECT_NAME}")
    _run_tests_project --> coverage_report("${Coverage_NAME}_coverage_report") 
```

knorth55 commented:
I checked this PR is working for these repos.
- gitai_druid
  - to check if the original issue is solved
- gitai_tools
  - to check if the original issue is solved
- gitai_filters
  - to check no test existing repo
- gitai_motor_tools
  - to check no test existing repo
